enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,294 @@
+---
+title: add-ImagePolicy-to-config.openshift.io-v1alpha1
+authors:
+  - "@QiWang19"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@saschagrunert, for API design, implementation details and graduation criteria"
+  - "@mrunalp"
+  - "@yuqi-zhang"
+  - "@mtrmac"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - "@mrunalp"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@mrunalp"
+creation-date: 2023-05-17
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/OCPNODE-1628
+---
+
+# Add CRD ImagePolicy to config.openshift.io/v1alpha1
+
+## Summary
+
+This enhancement propose an ImagePolicy CRD to provide a way to set sigstore image signature verification configuration on Openshift.
+
+## Motivation
+
+### User Stories
+
+As an OpenShift user, I want to verify the container images signed using sigstore
+tools, so that I can utilize the increased security of my software supply chain.
+
+### Goals
+
+- The user can create an ImagePolicy instance specifying the images/repositories to be verified and their policy. The MCO will write the configuration for signature verification. Once this is done, CRI-O can verify the images/repositories.
+
+- MCO container runtime config controller watches ImagePolicy instance in different namespaces and merges the instances for each namespace into a single [containers-policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md) in a predefined `/path/to/policies/\<NAMESPACE\>.json`.
+
+- CRI-O can verify Cosign signature signed images using configuration from ImagePolicy by using the namespace from the sandbox config on the `PullImage` RPC.
```

Comment:
⚠️ 

AFAICS that doesn’t work; the Kubelet does not trigger a `PullImage` if it already exists (depending on `Container.ImagePullPolicy`). So one permissively-configured namespace can cause an image to be pulled, and a strictly-configured namespace will not see its policy enforced.

This might trigger a fairly extensive rethink … does CRI-O need to trigger a pull on every `CreateContainer`? Somehow caching that outcome per namespace?

Reply:
@mtrmac would it be fine to call [`signature.IsRunningImageAllowed()`](https://github.com/containers/image/blob/8ad019310ff2ed8f99dea75b18998b415955bf8e/signature/policy_eval.go#L256-L293) on every container creation and caching the results?

Reply:
That directly works _if_ the image uses a digest reference. For tag references, we need to worry about pull getting image A, and container creation checking signatures of image B. I think it’s very likely code to do that safely can be written with the existing libraries, still I think prototyping that might be valuable to avoid late-stage surprises.

And MCO + CRI-O might need to cooperate about invalidating that cache if `policy.json` changes don’t always involve a reboot.

Reply:
Working towards a solution in https://github.com/cri-o/cri-o/pull/7046

Reply:
ISTM that this thread is really similar to https://github.com/containers/skopeo/issues/1829 right?

Reply:
@cgwalters I don’t see much of a connection. Here we are not worried about not having a policy, but about code paths not enforcing it (and possibly about just not having enough data to enforce it — @saschagrunert has fixed that part).

Reply:
My thought on the relation here was that I'd still like to consider persisting the result of the policy evaluation, because it makes it dramatically easier to introspect and verify.
We do this in ostree.

(Now yes, it introduces hysteresis because it raises the question of "how does it work if the policy changes", but I think caching e.g. the sha256 digest of the policy that was used at pull time would help in terms of verifying "did the policy change since the image was last pulled" etc.)